gobblin-runtime/src/main/java/gobblin/runtime/JobContext.java

```diff
@@ -332,7 +373,14 @@ void commit() throws IOException {
         allDatasetsCommit = false;
       } finally {
         try {
-          persistDatasetState(datasetUrn, datasetState);
+          if (commitSequence.isPresent()) {
+            commitSequence.get().append(buildDatasetStateCommitStep(datasetUrn, datasetState).get());
```

Comment:
Seems this is the only use case of `append`, and it seems `execute` only happens here after. So I think you can keep using the builder here instead of appending to a `CommitSequence`.